### ISSUE TITLE
YuE2: bound sampling to the legal-token window instead of scanning the full 184k-token vocab

### ISSUE THREAD

**Issue by T69** [scragnog/HOT-Step-CPP#169]
## Summary

`yue2_distribution`'s mask/temperature/top-k/top-p loops scanned the full `[0, V)` vocab (184,704 tokens) every sampling call, even though every index outside `[scan_lo, scan_hi)` — the legal range plus wherever `eos_id` sits relative to it — gets masked to `-inf` regardless of its value. This mirrors ACE-Step's own established pattern (`pipeline-lm.cpp`/`sampling.h`: compact logits to EOS + audio codes instead of the full 150k+ vocab).

Two bulk `-inf` fills replace the branchy full-vocab loop for the masking pass, and every later pass (temperature, top-k-apply, top-p-collect) only ever revisits `[scan_lo, scan_hi)` — nothing outside it can be finite again once the mask has run once.

`yue2_sample_argmax`/`yue2_sample_draw` are deliberately left untouched (full-V scan) since narrowing those would require signature changes touching `yue2-pipeline.h` and the `yue2-probe.cpp` parity tool — not worth the added risk for an already-cheap pass.

## Verification

Bit-exact: identical plan/semantic token hashes before and after, same seed.

## Measured (M1 Max, real production job)

Smaller than a back-of-envelope estimate suggested, but real and safe:
- semantic host-gap: 5724.4ms → 4881.1ms (**-843ms**)
- plan host-gap: 3266.8ms → 3082.8ms (**-184ms**)

## Note

Fully independent — applies cleanly on its own, no dependency on any other patch in this batch.

## Patch

​```diff
diff --git a/engine/src/yue2/yue2-sample.h b/engine/src/yue2/yue2-sample.h
index 8b602e66..089a8647 100644
--- a/engine/src/yue2/yue2-sample.h
+++ b/engine/src/yue2/yue2-sample.h
@@ -162,8 +162,30 @@ static inline void yue2_distribution(std::vector<float> & scores, const Yue2Samp
                                      int64_t step, bool legacy_off) {
     const int64_t V = (int64_t) scores.size();
 
+    // [scan_lo, scan_hi) bounds every index this call can ever leave finite:
+    // the legal range plus eos_id, wherever it sits relative to that range
+    // (adjacent for the semantic stage, ~200 ids past it for the abc stage --
+    // see the two real call sites in yue2-pipeline.h). Outside it, EVERY
+    // index is masked below regardless of its value, so two bulk -inf fills
+    // replace a 184704-entry branchy loop there, and every later pass in
+    // this function (temperature/top-k-apply/top-p-collect) only ever needs
+    // to revisit this same span: nothing outside it can be finite again once
+    // the mask below has run once. Semantics are unchanged -- this only
+    // changes which indices get touched by a branch vs. a bulk fill.
+    int64_t scan_lo = std::max<int64_t>(0, legal_lo);
+    int64_t scan_hi = std::min<int64_t>(V, legal_hi);
+    if (eos_id >= 0 && eos_id < V) {
+        scan_lo = std::min(scan_lo, eos_id);
+        scan_hi = std::max(scan_hi, eos_id + 1);
+    }
+    if (scan_lo > scan_hi) {
+        scan_lo = scan_hi;  // degenerate: empty legal range and no eos_id in range
+    }
+
     // legal-token mask (additive, TRUE -inf) + EOS whitelist + min_tokens re-block.
-    for (int64_t v = 0; v < V; v++) {
+    std::fill(scores.begin(), scores.begin() + scan_lo, -INFINITY);
+    std::fill(scores.begin() + scan_hi, scores.end(), -INFINITY);
+    for (int64_t v = scan_lo; v < scan_hi; v++) {
         const bool legal = (v >= legal_lo && v < legal_hi) || v == eos_id;
         if (!legal) {
             scores[(size_t) v] = -INFINITY;
@@ -202,7 +224,7 @@ static inline void yue2_distribution(std::vector<float> & scores, const Yue2Samp
         return;  // greedy short-circuit -- caller does an unfiltered argmax; top-k/top-p never run
     }
     if (sp.temperature != 1.0f) {
-        for (int64_t v = 0; v < V; v++) {
+        for (int64_t v = scan_lo; v < scan_hi; v++) {
             float & s = scores[(size_t) v];
             if (std::isfinite(s)) {
                 s = yue2_maybe_bf16(s / sp.temperature, legacy_off);
@@ -232,7 +254,7 @@ static inline void yue2_distribution(std::vector<float> & scores, const Yue2Samp
         }
     }
     if (std::isfinite(threshold)) {
-        for (int64_t v = 0; v < V; v++) {
+        for (int64_t v = scan_lo; v < scan_hi; v++) {
             if (scores[(size_t) v] < threshold) {
                 scores[(size_t) v] = -INFINITY;
             }
@@ -247,8 +269,8 @@ static inline void yue2_distribution(std::vector<float> & scores, const Yue2Samp
         // Sorting them wastes nearly all of the work for a 184,704-token
         // vocabulary when top-k leaves only a small set of candidates.
         std::vector<int64_t> idx;
-        idx.reserve((size_t) (sp.top_k > 0 ? std::min<int64_t>(sp.top_k, V) : V));
-        for (int64_t v = 0; v < V; v++) {
+        idx.reserve((size_t) (sp.top_k > 0 ? std::min<int64_t>(sp.top_k, V) : (scan_hi - scan_lo)));
+        for (int64_t v = scan_lo; v < scan_hi; v++) {
             if (std::isfinite(scores[(size_t) v])) idx.push_back(v);
         }
         if (idx.empty()) return;
​```

**Comment by scragnog** [comment]
Not merged, for the same underlying reason as #167, and I want to be precise about it because the patch is correct — it just no longer buys anything here.

`yue2_distribution` does not see a 184,704-wide vector on this fork. It takes a `base` argument and the caller passes only the head-window rows the stage samples from:

```
//   base   vocab id of scores[0] (doc 30 #1: the engine returns only the
//          lm_head rows a stage samples from). eos_id/legal_lo/legal_hi/
//          history stay ABSOLUTE ids; this function shifts them.
```

`const int64_t V = (int64_t) scores.size();` is therefore already `head_n`, and `eos_id -= base` runs before the point your patch inserts its bounds, so your `scan_lo`/`scan_hi` arithmetic would be consistent — it applies cleanly and it is not wrong. But for the semantic stage the legal range is essentially the whole of that narrowed window, so `scan_lo` lands near 0 and `scan_hi` near `V`, and the two bulk fills replace almost nothing. Net effect is a little extra branching.

Your -843ms on the semantic host-gap was measured against a build where the full vocabulary really did reach the sampler, so the number was true when you took it. On current master the saving it represents has already been taken at the source, by not producing those logits in the first place.

If you re-measure on master and there is still most of a second in there, please reopen — that would mean the head window is not doing what its call sites say it is, which I would want to know.

**Comment by scragnog** [comment]
Closing as already implemented — see the reply above. Reopen if re-measuring on current master still finds most of a second in the sampling pass.
